APDS9960 (my second try)

Staars opened the pull request:
This driver supports the APDS9960 in TASMOTA based on the recent code base (developed on 6.0a). The code is a stripped down and slightly modified version of the driver from Shawn Hymel/Sparkfun, who did the heavy work.

The sensor supports several modes (gestures, proximity, light level with color channels) and the driver should cover most of the possibilities. In it’s current form, it is intended to be used as a „user defined driver and sensor“ according to the API. Because of the quite big code size and its somehow exotical nature it is disabled by default. 
You must uncomment: #define USE_APDS9960 in one of the first lines to activate it (and to confirm, that you know, what you are doing ;). There are two (known) conflicting drivers (SHT and VEML6070) which will be deactivated including a compiler warning.

Connect the wires and configure SDA and SDL for I2C in Tasmota and the sensor should start in „gesture mode“.

With „rules“ in mind the gestures generate tele-mqtt-messages with the payload in the following scheme:

„APDS9960“:{„Up“:1}       for UP
„APDS9960“:{„Right“:1}   for RIGHT
„APDS9960“:{„Down“:1}  for DOWN
„APDS9960“:{„Left“:1}     for LEFT
„APDS9960“:{„Long“:1}   for LONG
„APDS9960“:{„None“:1}  for NONE

These can be used for rules too:
rule1 on Tele-APDS9960#Up do command endon

The LONG gesture will trigger after leaving the hand for a certain amount of time above the sensor and was initially introduced, to prevent a crash, because without it, the run loop will be blocked for ever. But it’s usable as a quite stable gesture (long press or long hold) and after it, the sensor will be switched on and off (and report this via mqtt).

NONE is simply a not recognised gesture, but shows, that there is some movement in front of the sensor.

With the command GESTURE you can switch between two modes:
gesture 1  -> read gestures
gesture 0 -> read light level and proximity (and report as TELE-message - also usable for rules).
gesture    -> get gesture mode (On or OFF, see both lines above)

In the non-gesture mode the JSON includes Red, Green, Blue and Proximity as raw sensor values (search for the data sheet of the APDS9960) and the ambient value should be Lux.

Please keep in mind, that there are hardware limits to the sensor (which is available for under 2 €) and the necessary and quite expensive computations in the run loop are fore sure not ideal for an environment like TASMOTA. 
Some sources in the internet suggest to use a small capacitor between VCC and Ground (I use 100µF), but maybe you come away without one.

I want to thank member Renne for his valuable feedback and hope, this is useful for sombody. 
If the maintainers do not intend to include this in the main project, it would be nice to get a response. In this case I will put the driver in small repository on its own.

P.S.:
I restart this PR because I accidentally deleted my old fork. I will close the old PR.

Staars committed:
Create xdrv_91_apds9960.ino

First version with the "new" driver model.
The driver is disabled by default to prevent side effects und must be enabled for testing by hand. 
Please uncomment:
// #define USE_APDS9960 

Use GPIO_DSB as interrupt pin as provisional hack in the web interface of tasmota.

Please expect problems (reboot, ...). My personal usage scenario is:
Sonoff -> MQTT -> Node-Red -> MQTT -> Sonoff

Staars committed:
Update xdrv_91_apds9960.ino

changes:
-changed from interrupt mode to polling
-fixed the mqtt-messages
-fixed command "gesture" for the console. Now every argument except 0 and 1 will report the gesture mode
-changed gesture name from "overload" to "long"

Staars committed:
Update xdrv_91_apds9960.ino

new:
-removed NEAR/FAR because of the unreliability
-support for Rules, therefore change of the MQTT-Messages to a number format (clockwise: up,right,down,left then long and none)
Example:
rule1 on Tele-APDS9960#Gesture=1 do ... endon

-minor code changes to shorten things

Staars committed:
Update xdrv_91_apds9960.ino

-add compiler warning (based on input from renne)
-small cosmetic/formatting cleanups

Staars committed:
Update xdrv_91_apds9960.ino

Changed the payload back to string.
I think this is shorter, easier to understand without any documentation and still works with rules.
e.g.
on Tele-APDS9960#Up do command endon

Staars commented:
Updated the Payload for the gestures back to string (after hopefully better understanding the mechanics of rules).
The messages are human readable now and this should be more convenient for applications like Node-Red or rules on the device, because there is nearly no need for documentation anymore.

renne commented:
The rule `on Tele-APDS9960#Ambient<100 do power on endon on Tele-APDS9960#Ambient>1000 do power off endon` powers the light on when brightness < 100 lux and powers light off when brightness > 1000 lux.

Staars commented:
Please try:
on Tele-APDS9960#Long do power toggle endon

I have shortened the syntax and dropped the "gesture" from the JSON.

renne commented:
`rule on Tele-APDS9960#Long do power toggle endon on Tele-APDS9960#Up do dimmer + endon on Tele-APDS9960#Down do dimmer - endon on Tele-APDS9960#Left do ct + endon on Tele-APDS9960#Right do ct - endon`

controls ON/OFF, brightness and color temperature

```
LONG:  toggle ON/OFF
UP:    increase brightness
DOWN:  decrease brightness
Left:  increase color temperature
Right: decrease color temperature
```

andrethomas commented:
Looks good, ordering myself some sensors!

arendst commented:
Will do some re-writing and move to xsns 27

Jason2866 commented:
Sensors looks promising, ordered some.
With integration in Tasmota nice "things" seems possible!
Thank you!

Staars commented:
Yes, please do.

Best regards,
Christian


> Am 26.06.2018 um 18:14 schrieb Jason2866 <notifications@github.com>:
> 
> Sensors looks promising, ordered some.
> With integration in Tasmota nice "things" seems possible!
> Thank you!
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

Jason2866 commented:
Hi @Staars 
Command: Gesture 0 /1 "Unknown"
And Sensors disappears from Web Frontend. Do you have a hint?
Latest Dev v.6.1.1c from today with arduino 2.3.0 

16:42:33 MQT: tele/sonoff-1585B3/LWT = Online (retained)
16:42:33 MQT: cmnd/sonoff-1585B3/POWER = 
16:42:33 MQT: tele/sonoff-1585B3/INFO1 = {"Module":"Generic","Version":"6.1.1c","FallbackTopic":"sonoff-1585B3","GroupTopic":"sonoffs"}
16:42:33 MQT: tele/sonoff-1585B3/INFO2 = {"WebServerMode":"Admin","Hostname":"sonoff-1585B3-1459","IPAddress":"192.168.2.149"}
16:42:33 MQT: tele/sonoff-1585B3/INFO3 = {"RestartReason":"Software/System restart"}
16:42:34 UPP: Multicast (re)joined
16:42:35 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
16:42:37 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:37","APDS9960":{"Long":1}}
16:42:37 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
16:42:40 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
16:42:41 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:41","APDS9960":{"Long":1}}
16:42:41 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
16:42:44 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
16:42:46 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:46","APDS9960":{"Long":1}}
16:42:46 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
16:42:47 MQT: tele/sonoff-1585B3/STATE = {"Time":"2018-07-28T16:42:47","Uptime":"0T00:00:22","Vcc":2.784,"Wifi":{"AP":1,"SSId":"Jason_Home_WLAN","RSSI":86,"APMac":"00:A0:57:2A:BD:19"}}
16:42:47 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:47","APDS99609":{"Red":0,"Green":0,"Blue":0,"Ambient":0,"Proximity":0}}
16:42:48 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
16:42:53 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:53","APDS99609":{"Right":1}}
16:43:57 CMD: Gesture 0
16:43:57 MQT: stat/sonoff-1585B3/RESULT = {"Command":"Unknown"}
16:44:03 CMD: Gesture 1
16:44:03 MQT: stat/sonoff-1585B3/RESULT = {"Command":"Unknown"}

Staars commented:
Hi,

just for my understandig:
Sensor works for gestures, but you do not get the values for rgbap in the web frontend, right? Do you use rules?

> Am 28.07.2018 um 16:47 schrieb Jason2866 <notifications@github.com>:
> 
> Hi @Staars
> Command: Gesture 0 /1 "Unknown"
> And Sensors disappears from Web Frontend. Do you have a hint?
> Latest Dev v.6.1.1c from today with arduino 2.3.0
> 
> 16:42:33 MQT: tele/sonoff-1585B3/LWT = Online (retained)
> 16:42:33 MQT: cmnd/sonoff-1585B3/POWER =
> 16:42:33 MQT: tele/sonoff-1585B3/INFO1 = {"Module":"Generic","Version":"6.1.1c","FallbackTopic":"sonoff-1585B3","GroupTopic":"sonoffs"}
> 16:42:33 MQT: tele/sonoff-1585B3/INFO2 = {"WebServerMode":"Admin","Hostname":"sonoff-1585B3-1459","IPAddress":"192.168.2.149"}
> 16:42:33 MQT: tele/sonoff-1585B3/INFO3 = {"RestartReason":"Software/System restart"}
> 16:42:34 UPP: Multicast (re)joined
> 16:42:35 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 16:42:37 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:37","APDS9960":{"Long":1}}
> 16:42:37 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 16:42:40 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 16:42:41 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:41","APDS9960":{"Long":1}}
> 16:42:41 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 16:42:44 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 16:42:46 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:46","APDS9960":{"Long":1}}
> 16:42:46 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 16:42:47 MQT: tele/sonoff-1585B3/STATE = {"Time":"2018-07-28T16:42:47","Uptime":"0T00:00:22","Vcc":2.784,"Wifi":{"AP":1,"SSId":"Jason_Home_WLAN","RSSI":86,"APMac":"00:A0:57:2A:BD:19"}}
> 16:42:47 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:47","APDS99609":{"Red":0,"Green":0,"Blue":0,"Ambient":0,"Proximity":0}}
> 16:42:48 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 16:42:53 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T16:42:53","APDS99609":{"Right":1}}
> 16:43:57 CMD: Gesture 0
> 16:43:57 MQT: stat/sonoff-1585B3/RESULT = {"Command":"Unknown"}
> 16:44:03 CMD: Gesture 1
> 16:44:03 MQT: stat/sonoff-1585B3/RESULT = {"Command":"Unknown"}
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Jason2866 commented:
@Staars 
No sensors works not for gestures. Sometimes it catches a gesture. Most of the time it doesnt work in general.

Jason2866 commented:
After restart:
![image](https://user-images.githubusercontent.com/24528715/43358247-ae9948dc-928e-11e8-8898-f777064b1acf.png)

and after a view seconds
![image](https://user-images.githubusercontent.com/24528715/43358251-c2140802-928e-11e8-8136-20009eabfa74.png)

Jason2866 commented:
This happens on webconsole

00:00:06 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
00:00:08 WIF: Connected
00:00:08 HTP: Web server active on sonoff-1585B3-1459 with IP address 192.168.2.149
00:00:08 UPP: Multicast (re)joined
17:49:51 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:49:53 RSL: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:49:53","APDS9960":{"Long":1}}
17:49:53 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:49:53 MQT: Attempting connection...
17:49:53 MQT: Connected
17:49:53 MQT: tele/sonoff-1585B3/LWT = Online (retained)
17:49:54 MQT: cmnd/sonoff-1585B3/POWER = 
17:49:54 MQT: tele/sonoff-1585B3/INFO1 = {"Module":"Generic","Version":"6.1.1c","FallbackTopic":"sonoff-1585B3","GroupTopic":"sonoffs"}
17:49:54 MQT: tele/sonoff-1585B3/INFO2 = {"WebServerMode":"Admin","Hostname":"sonoff-1585B3-1459","IPAddress":"192.168.2.149"}
17:49:54 MQT: tele/sonoff-1585B3/INFO3 = {"RestartReason":"Software/System restart"}
17:49:55 UPP: Multicast (re)joined
17:49:56 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:49:58 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:49:58","APDS9960":{"Long":1}}
17:49:58 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:00 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:02 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:02","APDS9960":{"Long":1}}
17:50:02 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:05 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:07 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:07","APDS9960":{"Long":1}}
17:50:07 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:08 MQT: tele/sonoff-1585B3/STATE = {"Time":"2018-07-28T17:50:08","Uptime":"0T00:00:22","Vcc":2.800,"Wifi":{"AP":1,"SSId":"Jason_Home_WLAN","RSSI":86,"APMac":"00:A0:57:2A:BD:19"}}
17:50:08 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:08","APDS99609":{"Red":0,"Green":0,"Blue":0,"Ambient":0,"Proximity":0}}
17:50:09 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:11 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:11","APDS99609":{"Long":1}}
17:50:11 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:14 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:16 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:16","APDS99609":{"Long":1}}
17:50:16 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:18 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:20 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:20","APDS99609":{"Long":1}}
17:50:20 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:23 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:24 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:24","APDS99609":{"Long":1}}
17:50:25 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:27 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:29 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:29","APDS99609":{"Long":1}}
17:50:29 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:32 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:33 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:33","APDS99609":{"Long":1}}
17:50:33 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:36 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:38 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:38","APDS99609":{"Long":1}}
17:50:38 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:40 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:42 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:42","APDS99609":{"Long":1}}
17:50:42 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:50:45 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:50:52 RSL: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:47","APDS99609":{"Long":1}}
17:50:57 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:51:08 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:51:10 RSL: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:10","APDS99609":{"Long":1}}
17:51:10 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:51:10 MQT: Attempting connection...
17:51:10 MQT: Connected
17:51:10 MQT: tele/sonoff-1585B3/LWT = Online (retained)
17:51:10 MQT: cmnd/sonoff-1585B3/POWER = 
17:51:11 UPP: Multicast (re)joined
17:51:13 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:51:15 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:15","APDS99609":{"Long":1}}
17:51:15 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:51:17 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:51:19 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:19","APDS99609":{"Long":1}}
17:51:19 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:51:22 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:51:23 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:23","APDS99609":{"Long":1}}
17:51:23 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:51:26 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:51:28 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:28","APDS99609":{"Long":1}}
17:51:28 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:51:30 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:51:32 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:32","APDS99609":{"Long":1}}
17:51:32 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
17:51:35 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
17:51:37 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:37","APDS99609":{"Long":1}}
17:51:37 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}

Staars commented:
Okay, the sensor sees something in front of it and reports a „long“ gesture. Then it switches off and on again, as intended.

As I wrote somewhere, that was the escape mechanism to protect the run loop of Tasmota, when I realized, that I can crash the esp8266 with Tasmota by just holding my hand in front of the sensor for some seconds. 

To safely debug this problem, you must rule out a „hardware“ problem first -> free sight for the sensor. To be really sure, you should connect the sensor to an Arduino and run a basic sketch of you choice.

In my tests on 3 different esp-devices the sensor was „a brave citizen“, but once the cables of the breadboard led to the effect, that you reported.


> Am 28.07.2018 um 17:52 schrieb Jason2866 <notifications@github.com>:
> 
> This happens on webconsole
> 
> 00:00:06 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 00:00:08 WIF: Connected
> 00:00:08 HTP: Web server active on sonoff-1585B3-1459 with IP address 192.168.2.149
> 00:00:08 UPP: Multicast (re)joined
> 17:49:51 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:49:53 RSL: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:49:53","APDS9960":{"Long":1}}
> 17:49:53 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:49:53 MQT: Attempting connection...
> 17:49:53 MQT: Connected
> 17:49:53 MQT: tele/sonoff-1585B3/LWT = Online (retained)
> 17:49:54 MQT: cmnd/sonoff-1585B3/POWER =
> 17:49:54 MQT: tele/sonoff-1585B3/INFO1 = {"Module":"Generic","Version":"6.1.1c","FallbackTopic":"sonoff-1585B3","GroupTopic":"sonoffs"}
> 17:49:54 MQT: tele/sonoff-1585B3/INFO2 = {"WebServerMode":"Admin","Hostname":"sonoff-1585B3-1459","IPAddress":"192.168.2.149"}
> 17:49:54 MQT: tele/sonoff-1585B3/INFO3 = {"RestartReason":"Software/System restart"}
> 17:49:55 UPP: Multicast (re)joined
> 17:49:56 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:49:58 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:49:58","APDS9960":{"Long":1}}
> 17:49:58 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:00 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:02 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:02","APDS9960":{"Long":1}}
> 17:50:02 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:05 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:07 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:07","APDS9960":{"Long":1}}
> 17:50:07 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:08 MQT: tele/sonoff-1585B3/STATE = {"Time":"2018-07-28T17:50:08","Uptime":"0T00:00:22","Vcc":2.800,"Wifi":{"AP":1,"SSId":"Jason_Home_WLAN","RSSI":86,"APMac":"00:A0:57:2A:BD:19"}}
> 17:50:08 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:08","APDS99609":{"Red":0,"Green":0,"Blue":0,"Ambient":0,"Proximity":0}}
> 17:50:09 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:11 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:11","APDS99609":{"Long":1}}
> 17:50:11 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:14 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:16 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:16","APDS99609":{"Long":1}}
> 17:50:16 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:18 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:20 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:20","APDS99609":{"Long":1}}
> 17:50:20 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:23 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:24 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:24","APDS99609":{"Long":1}}
> 17:50:25 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:27 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:29 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:29","APDS99609":{"Long":1}}
> 17:50:29 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:32 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:33 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:33","APDS99609":{"Long":1}}
> 17:50:33 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:36 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:38 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:38","APDS99609":{"Long":1}}
> 17:50:38 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:40 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:42 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:42","APDS99609":{"Long":1}}
> 17:50:42 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:50:45 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:50:52 RSL: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:50:47","APDS99609":{"Long":1}}
> 17:50:57 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:51:08 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:51:10 RSL: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:10","APDS99609":{"Long":1}}
> 17:51:10 RSL: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:51:10 MQT: Attempting connection...
> 17:51:10 MQT: Connected
> 17:51:10 MQT: tele/sonoff-1585B3/LWT = Online (retained)
> 17:51:10 MQT: cmnd/sonoff-1585B3/POWER =
> 17:51:11 UPP: Multicast (re)joined
> 17:51:13 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:51:15 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:15","APDS99609":{"Long":1}}
> 17:51:15 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:51:17 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:51:19 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:19","APDS99609":{"Long":1}}
> 17:51:19 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:51:22 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:51:23 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:23","APDS99609":{"Long":1}}
> 17:51:23 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:51:26 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:51:28 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:28","APDS99609":{"Long":1}}
> 17:51:28 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:51:30 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:51:32 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:32","APDS99609":{"Long":1}}
> 17:51:32 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 17:51:35 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"On"}
> 17:51:37 MQT: tele/sonoff-1585B3/SENSOR = {"Time":"2018-07-28T17:51:37","APDS99609":{"Long":1}}
> 17:51:37 MQT: tele/sonoff-1585B3/RESULT = {"Gesture":"Off"}
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/arendst/Sonoff-Tasmota/pull/3051#issuecomment-408616938>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AFOiZHJMJLv3DFi8TTa9A13K-gt_9phzks5uLIi7gaJpZM4U0wbw>.
>

Jason2866 commented:
Ok will try with a arduino...
But it is connected to a WeMos mini via short cables, no breadboard (i had bad experiences too!)
and it is nothing in front off. The sensor is the only device connected to the WeMos.

Jason2866 commented:
Will try the second i have ordered, maybe this one is a "brave one"

Jason2866 commented:
But why is Gesture 0 not working?
19:11:34 CMD: Gesture 0
19:11:34 MQT: stat/sonoff-1585B3/RESULT = {"Command":"Unknown"}

Jason2866 commented:
Nothing relevant in log. Thank you for trying to help me!

Staars commented:
Okay, it seems that not all of my code did survive the merging process ;)

I can not use the gesture command at the moment, but I am optimistic, that we can solve this. I am not at home at moment, maybe I find the time today in the evening.

But at least the gesture mode works for me with 6.1.1c. No issues here. I just cannot switch between the modes (gesture <> RGBAP).

You must see a compiler warning (deactivating SHT and VEML6070), otherwise there is no APDS9960 and it is deactivated by default.

> Am 28.07.2018 um 20:38 schrieb Jason2866 <notifications@github.com>:
> 
> Nothing relevant in log. Thank you for trying to help me!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/arendst/Sonoff-Tasmota/pull/3051#issuecomment-408627133>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AFOiZIa_bE74aSHKcaMoVeCt3VzSHa1Oks5uLK-sgaJpZM4U0wbw>.
>

Jason2866 commented:
I have deactivated SHT and VEML6070 in user_config_override.h (undef...)
I see no compiler warning. I will play around with this setting, but i think this will not change things.
No hurry for bug tracking, thanks for taking a look :-)

Staars commented:
Of course you can deactivate SHT and VEML manually, but this would happen automatically when you activate APDS9960 including a compiler warning. The (very tiny) advantage is that you can be 101% sure, that you really have selected APDS9960 for compiling. This is only for convenience and you can handle it as you like.

My bug hunting activities will be closely related to the insane air temperatures in the next days ;)

> Am 29.07.2018 um 18:39 schrieb Jason2866 <notifications@github.com>:
> 
> I have deactivated SHT and VEML6070 in user_config_override.h (undef...)
> I see no compiler warning. I will play around with this setting, but i think this will not change things.
> No hurry for bug tracking, thanks for taking a look :-)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Staars commented:
Okay, now I got it.

At the moment you have to use SENSOR27 (0|1) instead of GESTURE (0|1). Maybe Theo is in the middle of refactoring some things … we will see.

So for me there is nothing wrong with the current code base. 

My APDS9960 is fully functional in Tasmota and I fear, you have to check your sensor with an Arduino or something similiar. If you do not use a small capacitor (over vcc and ground), try this as the TRM mentions electrical noise, which influences proximity readings (even if the sensor works with an Arduino without problems!). Normal proximity raw value ist under 10 and it increases to max. 255, when something is very close to the sensor.

I hope you get it to work!

> Am 29.07.2018 um 18:39 schrieb Jason2866 <notifications@github.com>:
> 
> I have deactivated SHT and VEML6070 in user_config_override.h (undef...)
> I see no compiler warning. I will play around with this setting, but i think this will not change things.
> No hurry for bug tracking, thanks for taking a look :-)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/arendst/Sonoff-Tasmota/pull/3051#issuecomment-408689732>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AFOiZPO40oYX4uOptxujmzrZR4aEbqqQks5uLeVYgaJpZM4U0wbw>.
>

Jason2866 commented:
Thx for taking a look.
Will solder tomorrow capacitor(s) 1nF and 100uF. Will see...

Jason2866 commented:
@andrethomas 
Thx. Will take these.  
My choice was the experience from the days (long time ago...) building electronic "things" by my own
with discrete components and low level ic´s (74HCT...)

andrethomas commented:
@Jason2866 Yeah, the good old days when every single component had to be decoupled with 100nF capacitors :) A 1uF usually provides good filtering for AC level noise - I guess they call for >= 10uF (so 100uF as you said would also be fine) is probably just to filter our transient voltages generated by the led pulses. For really sensitive circuits such as high bit ADC's I usually do 100nF, 1uF and 10uF.

Jason2866 commented:
Soldered Tantal Elko...  Now is "totally silence"
`11:09:20 MQT: tele/sonoff-1585B3/LWT = Online (retained)
11:09:20 MQT: cmnd/sonoff-1585B3/POWER = 
11:09:20 MQT: tele/sonoff-1585B3/INFO1 = {"Module":"Generic","Version":"6.1.1c","FallbackTopic":"sonoff-1585B3","GroupTopic":"sonoffs"}
11:09:20 MQT: tele/sonoff-1585B3/INFO2 = {"WebServerMode":"Admin","Hostname":"sonoff-1585B3-1459","IPAddress":"192.168.2.149"}
11:09:20 MQT: tele/sonoff-1585B3/INFO3 = {"RestartReason":"Software/System restart"}
11:09:28 MQT: tele/sonoff-1585B3/STATE = {"Time":"2018-07-30T11:09:28","Uptime":"0T00:00:15","Vcc":2.783,"Wifi":{"AP":1,"SSId":"Jason_Home_WLAN","RSSI":82,"APMac":"00:A0:57:2A:BD:19"}}
11:09:35 CMD: Sensor27 0
11:09:35 MQT: stat/sonoff-1585B3/RESULT = {"Sensor27":"OFF"}`

![image](https://user-images.githubusercontent.com/24528715/43388375-3390da2c-93e9-11e8-83cf-11f65f44bc5c.png)

Will connect to a 3.3V Arduino mini and do testing. To be continued...

Jason2866 commented:
The sensor isnt working under Arduino too. None of the examples is working
`failed to initialize device! Please check your wiring.`
I have it connected to a ProMini 3.3V (ATmega328P) SCL on Pin A5 and SDA on Pin A4

At the moment no idea This module i have bought:
https://www.aliexpress.com/item/Free-Shipping-GY-9960-3-3-APDS-9960-proximity-detection-and-non-contact-gesture-detection-RGB/32708290585.html?spm=a2g0s.9042311.0.0.46fd4c4d6Cmhtl

andrethomas commented:
@Jason2866 

Try the sketch from here https://playground.arduino.cc/Main/I2cScanner onto the pro mini 3.3v and see if it at least finds the i2c address.

Jason2866 commented:
@andrethomas 

Ahh... On ESP it was found at 0x39

On Arduino Mini:
Scanning...
No I2C devices found

Staars commented:
I‘m pretty sure, that every Tasmota-screenshot from you proves, that a device with the correct adress (0x39) was found, because otherwise the web interface should not show up.

And while you are at testing on Arduino, please test an apds9930-sketch too. Just an idea ...

My conclusion so far is, that the sensor is found, but does not respond correctly.


> Am 30.07.2018 um 13:01 schrieb andrethomas <notifications@github.com>:
> 
> @Jason2866
> 
> Try the sketch from here https://playground.arduino.cc/Main/I2cScanner onto the pro mini 3.3v and see if it at least finds the i2c address.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Jason2866 commented:
@Staars 
I have tested on the arduino the example sketch(es) from Adafruit with the result 
`failed to initialize device! Please check your wiring.`

Jason2866 commented:
ok, i have a problem with arduino why? i dont know! 
Tried a other i2c device (MCP23017) which works fine under Tasmota -> i2c scanner didnt find it.

andrethomas commented:
Did you solder any of the bridges on the pc board - can you send a photo of both sides of your module's board?

Jason2866 commented:
![image](https://user-images.githubusercontent.com/24528715/43411051-5333ecd0-9429-11e8-9307-e99eac7fa5a2.png)
![image](https://user-images.githubusercontent.com/24528715/43411191-c34b87c6-9429-11e8-8b59-18fdcdd2c13f.png)

Jason2866 commented:
Now got it running under Arduino. Silly error! Did connect wrong to 4 and 5 not A4 and A5. Ahhhh!
- Gesture sketch works 
- Color sensor sketch works
- Proximity sketch says only device initialized (i think this is okay because int pin NOT connected!)

Staars commented:
> So i think it is a problem of the code in Tasmota

That is possible :(

The problem for me is, that my sensor works without a problem and that makes the debugging process really hard.
Maybe you can experiment with the default values (line 212-239), but beside that, I do not have good idea at the moment. 
At least my sensor has a different PCB than yours and maybe needs different values.

andrethomas commented:
Board looks good and jumpers for pull ups are soldered so it should at least be detected on I2Cscan. Which pins are you using for I2C on Tasmota? 

![image](https://user-images.githubusercontent.com/470015/43419463-3eb6a078-9441-11e8-9497-c10864a965e2.png)

Unfortunately, my module is still on a slow tugboat from China...

andrethomas commented:
Also, those caps you added are not necessary - can see the 10uF on the breakout board and right above it is probably a 100nF

Jason2866 commented:
@andrethomas 
Tried different GPIO pins.... Always the same. Sensor is found at 0x39. 
Sensor is correct identified because i get the Web Info of the sensor.
As @Staars wrote he thinks it is a problem in the code in combination with Tasmota.
The sensor works pretty well with arduino (after correcting my silly wiring fault).

Staars commented:
Does cold/warm reset(reboot) make a difference?

Jason2866 commented:
@andrethomas 
Will try without additional caps (tomorrow)...

andrethomas commented:
@Jason2866 Not that the caps would have a negative impact - just that they do not appear to be necessary so I don't think you should expect different results just yet.

Jason2866 commented:
@Staars 
No difference cold or warm boot. It never works with Tasmota

andrethomas commented:
Just to check, you running @ 80mhz right?

Jason2866 commented:
@andrethomas 
Without the additional caps Tasmota was crazy with on off switching gesture mode!
With caps it is totally silence!

andrethomas commented:
@Jason2866 Sure, leave them there then - can't do any harm but it is questionable why you'd get crazy data on I2C... tried a different esp module? Sorry... just running through the switches in my mind as things I would check if I actually had a board... if it's urgent I can probably get one DHL'd from banggood.

Jason2866 commented:
@andrethomas 
Yes 80 Mhz 
I use VSC and i changed only two settings in platformio.ini -> user_config_override.h and use core 2.3.0

Jason2866 commented:
@andrethomas 
BIG THX FOR HELP! No it is not urgent. It is a project that i like to do for me. When its done its done....
I can wait until you receive your sensors from china...
Tried already on two different Esp devices (wemos mini and NodeMCU). 
Will check on a Sonoff basic tomorrow. But i think the results will be negative too...

Jason2866 commented:
@andrethomas 
1) I2C without pullups? What i read it is neccersary?! But he, nothing to loose. Will try
2) Thats easy, will connect MCP23017 (did i say your driver works perfect :-) )

Jason2866 commented:
@Staars 
Arduino sketch is working with my sensor. Can i copy the initializing settings?
With which settings i can play around? What settings from your code differs to the arduinp code?

Staars commented:
I guess it was LED-boost from 300 to 100. Take a look here, as it seems the sensors on the market do really differ substantially: https://github.com/sparkfun/SparkFun_APDS-9960_Sensor_Arduino_Library/issues/9

Try to build the most basic example on the ESP to rule out hardware related problems.

BTW, I just ordered „your“ sensor, but this will take a while.


> Am 30.07.2018 um 22:54 schrieb Jason2866 <notifications@github.com>:
> 
> @Staars
> Arduino sketch is working with my sensor. Can i copy the initializing settings?
> With which settings i can play around? What settings from your code differs to the arduinp code?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Jason2866 commented:
@Staars 
Will try playing around with Led boost (tomorrow). I am on holiday sitting on my terrace with a glass of wine 😃
Thanks for ordering "my" sensor to get this issue solved.

Jason2866 commented:
Status:
Sensor is identified and I2C communication works.
I played around with some values (GGAIN_2X, setLEDBoost(LED_BOOST_300);  // back to 300)
but not really a success. SOMETIMES a gesture is identified.

What is different from this code https://learn.adafruit.com/adafruit-apds9960-breakout/arduino-wiring-and-test
and yours? The adafruit code works as a charm on the Arduino Pro Mini.

Jason2866 commented:
Tried https://github.com/sparkfun/SparkFun_APDS-9960_Sensor_Arduino_Library
With this library the sensor DOESNT work!
`--------------------------------
SparkFun APDS-9960 - GestureTest
--------------------------------
APDS-9960 initialization complete
Gesture sensor is now running

--------------------------------
SparkFun APDS-9960 - GestureTest
--------------------------------
APDS-9960 initialization complete
Gesture sensor is now running
`

Staars commented:
In a quick look at both libraries the main difference seems to be a power cycle of the chip (line 108 - 113 in the Adafruit version).
You can try to add this to the TASMOTA driver with:

disablePower();
delay(10);
enablePower();
delay(10);

Insert this at line 1235 (at least somewhere in the init function). 
Did you get the basic examples from the Adafruit-Library to work on the ESP (not the Arduino)?

Btw, i am not able to contribute anything in the next few hours. So expect a delay from my side.


> Am 31.07.2018 um 12:45 schrieb Jason2866 <notifications@github.com>:
> 
> Tried https://github.com/sparkfun/SparkFun_APDS-9960_Sensor_Arduino_Library
> With this library the sensor DOESNT work!
> `--------------------------------
> SparkFun APDS-9960 - GestureTest
> 
> APDS-9960 initialization complete
> Gesture sensor is now running
> 
> SparkFun APDS-9960 - GestureTest
> 
> APDS-9960 initialization complete
> Gesture sensor is now running
> `
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

renne commented:
@Jason2866 
Can you try compiling and uploading the code with the [Arduino IDE 1.8.5](https://github.com/arendst/Sonoff-Tasmota/wiki/Arduino-IDE) and ESP8266 Board version 2.4.1 to rule out IDE/Compiler issues?

Jason2866 commented:
@renne 
Uh, i really hate Arduino IDE for complex projects.
Will try to get it running. I never used it for Tasmota.

Jason2866 commented:
@renne 
I dont think it is a compiler issue because the original sketch / library from Sparkfun isnt working on Arduino board either! The Adafruit library / sketch IS working on Arduino.

Jason2866 commented:
@Staars 
Inserted as you suggest, didnt helped:

```
bool APDS9960_init()
{
    /* Set default values for ambient light and proximity registers */

    disablePower();

    delay(10);

    enablePower();

    delay(10);

    I2cWrite8(APDS9960_I2C_ADDR, APDS9960_ATIME, DEFAULT_ATIME)  ;

    I2cWrite8(APDS9960_I2C_ADDR, APDS9960_WTIME, DEFAULT_WTIME) ;

    I2cWrite8(APDS9960_I2C_ADDR, APDS9960_PPULSE, DEFAULT_PROX_PPULSE)  ;

    I2cWrite8(APDS9960_I2C_ADDR, APDS9960_POFFSET_UR, DEFAULT_POFFSET_UR)  ;

    I2cWrite8(APDS9960_I2C_ADDR, APDS9960_POFFSET_DL, DEFAULT_POFFSET_DL)  ;

    I2cWrite8(APDS9960_I2C_ADDR, APDS9960_CONFIG1, DEFAULT_CONFIG1)  ;

    setLEDDrive(DEFAULT_LDRIVE);
```

Staars commented:
Did the Adafruit sketch work on the ESP?


> Am 31.07.2018 um 13:46 schrieb Jason2866 <notifications@github.com>:
> 
> @Staars
> Inserted as you suggest, didnt helped:
> 
> bool APDS9960_init()
> {
>     /* Set default values for ambient light and proximity registers */
> 
>     disablePower();
> 
>     delay(10);
> 
>     enablePower();
> 
>     delay(10);
> 
>     I2cWrite8(APDS9960_I2C_ADDR, APDS9960_ATIME, DEFAULT_ATIME)  ;
> 
>     I2cWrite8(APDS9960_I2C_ADDR, APDS9960_WTIME, DEFAULT_WTIME) ;
> 
>     I2cWrite8(APDS9960_I2C_ADDR, APDS9960_PPULSE, DEFAULT_PROX_PPULSE)  ;
> 
>     I2cWrite8(APDS9960_I2C_ADDR, APDS9960_POFFSET_UR, DEFAULT_POFFSET_UR)  ;
> 
>     I2cWrite8(APDS9960_I2C_ADDR, APDS9960_POFFSET_DL, DEFAULT_POFFSET_DL)  ;
> 
>     I2cWrite8(APDS9960_I2C_ADDR, APDS9960_CONFIG1, DEFAULT_CONFIG1)  ;
> 
>     setLEDDrive(DEFAULT_LDRIVE);
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Jason2866 commented:
@Staars 
Which pins i have to use for SCL and SDA on a Wemos Mini with the original Adafruit sketch?

andrethomas commented:
I don't have a wemos mini yet but looks like D1->SCL and D2->SDA

Jason2866 commented:
@Staars @renne 
YES, original sketch "Gesture" from Adafruit is working on the Wemos mini 

```
Device initialized!
>
<
<
v
^
>
<
>

```

Jason2866 commented:
Sparkfun proximity and color sketch works on Wemos mini too. 
Gesture Sparkfun sketch fails

```
--------------------------------
SparkFun APDS-9960 - GestureTest
--------------------------------
APDS-9960 initialization complete
Gesture sensor is now running
```

andrethomas commented:
So the only known difference so far is the core version? It does not make sense that it would have an impact.

Jason2866 commented:
@andrethomas 
No i tried both core versions to exclude that error comes from there!

Adafruit gesture worked with both core versions on the ESP.
Sparkfun gesture failed with both core versions on the ESP.

Adafruit gesture worked on Arduino Pro mini.
Sparkfun gesture failed on Arduino Pro mini.

So for me the conclusion is that the Sparkfun code isnt working correct with the sensor in gesture mode!

renne commented:
The `F()` macros in MCP23017 worked with Platform.IO but caused memory access errors with Arduino IDE 1.8.5. It's far-fetched but maybe there is some general problem with the `F()` macros. Or just some initialization issue with the APDS-9960. I wasn't able to use RGBC and gesture mode at the same time, yet.

Jason2866 commented:
@renne 
Doesnt use Adafruit` F()` macros? Because there code is working.
The Sparkfun gesture never worked (not in Tasmota or the example sketch in arduino)...
I do compile Tasmota with VSC. Can try PlatformIO (Atom.io) too, pretty sure will not change result.
I think it is a initialization and "handling" problem. It looks like the sensors are different.
@Staars  posted a link where sparkfun users had problems with boards from sparkfun too

andrethomas commented:
@renne  `I wasn't able to use RGBC and gesture mode at the same time, yet.`

Is that so that it will know if your face is RED or GREEN whilst gesturing? :)

I find the F() and other PROGMEM related functions to be pretty random in when it works and when it doesn't. Since most of the functions just run quick in ram and would dump the content you could probably just remove all the F(), PTR(), and FPTR() from the code.

renne commented:
@andrethomas 
```
switch (face) {
  case red;
    disableHeating();
    break;
  case blue;
    enableHeating();
    break;
  case green;
    openWindow();
    break;
  default:
    beBored();
    break;
}  // :-)
```
Or just use gestures && RGBC for a closed-loop RGBCWWW ambilight  control with the H801. ;-)

Staars commented:
If things do not get sorted out until my new sensor arrives (the Jason-model;), I will try to rebase the driver (unless someone else will do it). But I will not do this without the needed hardware.
My conclusion so far is: The different models of the APDS9960 differ more, than I could imagine.


> Am 31.07.2018 um 15:30 schrieb renne <notifications@github.com>:
> 
> @andrethomas
> 
> switch (face) {
>   case red;
>     disableHeating();
>     break;
>   case blue;
>     enableHeating();
>     break;
>   case green;
>     openWindow();
>     break;
>   default:
>     beBored();
>     break;
> }  // :-)
> Or just use gestures && RGBC for a closed-loop RGBCWWW ambilight control with the H801. ;-)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

andrethomas commented:
@Staars @renne Is it possible to add support to this driver for the APDS9930 ? It uses the same address, but I am not sure how differential detection should be handled.

Staars commented:
I will try to implement this, but please understand, that I will wait for the „problematic“ sensor (that I ordered) before I return to the drawing board.
Meanwhile I will have a look at some implementations for the APDS9930. I am quite optimistic, that a sensor differentiation should be possible.

andrethomas commented:
Actually, looks like it will be

For 9930, read APDS9930_ID 0x12 will be = 0x12 or 0x39
For 9960 read APDS9960_ID 0x92 will be = 0xAB or 0x9C

I'm just wondering since I seem to have received APDS9930's instead of APDS9960's maybe validation could rule out bad sellers :|

No rush... was just curious - Added some APDS9960's to my banggood cart which will ship with DHL as soon as other items in my cart are back in stock so maybe 2 weeks or so

I will test the above theory and revert back if I can confirm my theory that I've been given APDS9930's from aliexpress.

andrethomas commented:
Ah, you do the right thing from the start... thats why mine was not detected LOL

`  if (APDS9960type == APDS9960_CHIPID_1 || APDS9960type == APDS9960_CHIPID_2) {`

andrethomas commented:
Perhaps since APDS9930 is older and has less function, let's not regress and add support for it. So leave as is :)

andrethomas commented:
```
21:56:36 MQT: stat/sonoff/RESULT = {"I2CScan":"Device(s) found at 0x39"}
21:56:36 MQT: stat/sonoff/RESULT = {"Device ID": = 18 }

```
So thats a 0x12 for me... confirmed to be 9930 and 9960.. damn chinese!

Staars commented:
Hello (again), I received the "jason"-model yesterday and began with coding.

The main "problem" is to really and securely reproduce the problem with the non-working sensor, because meanwhile I have updated my local repository and maybe there are side effects.

Long story short: The sensor works for me.

BTW, I have added calculated color temperature (from the Adafruit library) and will at least add a small check for the APDS9930 to give a hint, if the user purchased an incorrect sensor.

My todo-list:
Play a bit with sensor and code today (and check for stability!!).
Push the new things to my Tasmota-fork at GitHub (hopefully tomorrow).
Clean up and add some things in the next week.
Get some response about success or not from here.
Make a PR.

Jason2866 commented:
Hello @Staars 
if you have something to test i will find time this weekend 🤗👍

Staars commented:
Okay, I uploaded a new version to my Tasmota-fork.

The problem is, I am not totally shure about this "problematic" sensor. In my unscientific tests, it showed a bit more (but still very few) NONE-gestures and was more nitpicky regarding the super-correct placing of the sensor.
My best guess is, that it draws a tiny amount more power and our power budget is very tight on most ESP8266-devices (Tasmota shows 2,7 volt on my nodemcu). In clear sight my old sensor shows a proximity value of about 6-7. The problematic version shows mostly 20 and I had the feeling, that sometimes (after touching the wires on my breadboard) it was even higher. In this condition I got the LONG-gesture all the time. We can influence this with the gain factor in the code (which is 4 by default), but then I expect other problems (bad gesture readings). I think we see noise in the signal here, which might be power related, but I do not have the knowledge and equipment to really prove this.

Anyway, please try the new version, which should end up in a new PR.  I´m still experimenting with the color reading function, which might be wrong at the moment.

I hope we will make progress, because my new MGC3130-board is waiting in the corner 😏

andrethomas commented:
The drop in voltage reported may be due to the infrared led used by the sensors... perhaps that's why they make provision for a separate led power input (VL) ? Just thinking out loud here but I remember from using IR led's before that they tend to need more current to function properly...

Staars commented:
Yes, I agree. 
But I think we have a „boderline“ problem, so my first sensor works completely out of the box and the other one seems to be in the diffuse area of „works or works not“. (in german: ... auf Kante genäht;)
Next week I will try out some things on my Arduino. Maybe, it sheds some new light on this.


> Am 25.08.2018 um 17:58 schrieb andrethomas <notifications@github.com>:
> 
> The drop in voltage reported may be due to the infrared led used by the sensors... perhaps that's why they make provision for a separate led power input (VL) ? Just thinking out loud here but I remember from using IR led's before that they tend to need more current to function properly...
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

andrethomas commented:
flüchtig :)

Jason2866 commented:
@Staars 
Will try your latest fork tonight or tomorrow with my "Sorgenkind" device
Viele Grüße aus Bayern in dem es seit langem wieder einmal regnet
Servus Jason2866

Staars commented:
Yes, please give it a try. But do not expect miracles, there are not exactly specific tweaks for your/our problem in this new version.

Greetings from the german desert (aka Brandenburg).


> Am 25.08.2018 um 20:14 schrieb Jason2866 <notifications@github.com>:
> 
> @Staars
> Will try your latest fork tonight or tomorrow with my "Sorgenkind" device
> Viele Grüße aus Bayern in dem es seit langem wieder einmal regnet
> Servus Jason2866
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Jason2866 commented:
@Staars 
No change in behaviour. Still not working

Staars commented:
Yes, that is not unexpected :(

The problem is: Even if „your“ sensor model shows some little glitches in my setup, it is definitely usable and I fear it is impossible for me, to find a work around without the same hardware. At least I will do some tests with the different libraries (Sparkfun vs Adafruit) on my Arduino, but I probably will not find the time in the next few days.

Last questions:
Does your sensor not show up or do you see the „LONG-gesture cycle“? What is the proximity value with complete free sight (after SENSOR27 0)?
If your sensor does not show up, can you make a device reset (reset 0) and test again?
If you suffer from boredom, can you test your sensor with the Adafruit-library on your ESP (if you did not do this already, but I guess you tested it on the Arduino)?

Best regards



> Am 26.08.2018 um 12:43 schrieb Jason2866 <notifications@github.com>:
> 
> @Staars <https://github.com/Staars>
> No change in behaviour. Still not working
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/arendst/Sonoff-Tasmota/pull/3051#issuecomment-416029370>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AFOiZO_Wdc36RSPPwCuQ24Q71W0Uwdt6ks5uUnvjgaJpZM4U0wbw>.
>

Jason2866 commented:
I tested my 2 sensors under Atmega and ESP device with Adafruit. On both the worked perfect (see this thread). Sensor does not show up in Webconsole. It switches Gesture mode without interaction.

Jason2866 commented:
I can send you one if this helps.

Staars commented:
Yes, why not. 

Is there a private message system here on GitHub?

> Am 26.08.2018 um 20:37 schrieb Jason2866 <notifications@github.com>:
> 
> I can send you one if this helps.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/arendst/Sonoff-Tasmota/pull/3051#issuecomment-416059364>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AFOiZM5n_Ja_ywrum3JkMw_jjZBaNCw-ks5uUurTgaJpZM4U0wbw>.
>

Jason2866 commented:
Wegwerf eMail Adresse von mir jason2866@existiert.net 
Schreib mir deine Adresse und ich schick dir das bockige Teil ;-)

Jason2866 commented:
Habe deine Mail mit der Anschrift bekommen. Schau das ich Montag oder Dienstag auf die Post schaffe...
Danke schon einmal!

andrethomas commented:
Möglicherweise möchten sie das nach der fertigstellung löschen, sonst werden sie von einigen dummen spammed :|

Jason2866 commented:
@andrethomas 
Thank you for your comment. It is a junk eMail one time account. It self destructs in 1 hour... 😎

renne commented:
I've contacted the Broadcom/EBV support about more information on the APDS-9960 initialization values and color-space. Hopefully they'll answer soon.

Jason2866 commented:
Hallo Christian,
der Sensor ist auf dem Postweg.

Staars commented:
Hi, 

it has just arrived. Due to my job, I will probably not find enough testing-time before thursday. I will keep You updated.
Thanx!!


> Am 28.08.2018 um 15:27 schrieb Jason2866 <notifications@github.com>:
> 
> Hallo Christian,
> der Sensor ist auf dem Postweg.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Staars commented:
I just updated my fork.

After playing with the sensor of Jason2866 I added a command "sensor27 2" which will turn on the gesture mode with a limited gain factor for proximity and gesture (from 4x to 2x). That works for me, but I start to believe, that the APDS9960 (depending on the very model) can raise problems with other hardware combinations. So I probably have to make an entry in the wiki.

I added calculated illuminance in lux, which is quite close to the readings of a BH1750. The raw ambient/clear color values of the sensor do have a different meaning. I did not know this before.

@https://github.com/renne: 
Could you please check, that the driver is still working for you. It should change nothing for you out of the box.

andrethomas commented:
@Staars Is it by chance possible to measure current requirements to rule out sink rate issues from the IR led ?

Jason2866 commented:
@Staars 
Sensor27 2 works with my sensor :-) 👍 
AND i have found the conflicting driver why i doesnt worked with my usual setup.
If the TSL2561 driver is enabled the APDS9960 isnt working anymore :-(

So in general good news :-) 

If you or some can take a look into why this two drivers are conflicting
and solve this issue (if possible) would be perfect!

Jason2866 commented:
@Staars 
Could you do a PR for the new version?

Staars commented:
Sadly there is a lack of knowledge and equipment on my side. 

I only have a ultra-cheap multimeter here. Is that up to the task?

> Am 30.08.2018 um 16:10 schrieb andrethomas <notifications@github.com>:
> 
> @Staars <https://github.com/Staars> Is it by chance possible to measure current requirements to rule out sink rate issues from the IR led ?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/arendst/Sonoff-Tasmota/pull/3051#issuecomment-417333241>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AFOiZLjOyT8_T7o-6kv1GUz3PtLx81J9ks5uV_JrgaJpZM4U0wbw>.
>

Staars commented:
Hmm, after re-reading the data-sheet a few times, I think that the diagrams on page 7 strongly suggest, that the clear color value is in nearly linear correlation to a lux-meter.
Both libraries (Sparkfun and Adafruit) use a AGAIN-Value of 4 by default. So I tried 1 (=no light gain) instead and now I see values for the clear color channel that almost match the values of my BH1750 right beside. Of course now the computed lux values are far off, but I think that the "Adafruit-way" is wrong here. The data sheet only says "... can be computed..." but does not show how. This is the real root of confusion.

It would be nice, if someone can confirm this and if no one protests, I will do it that way now. Maybe we lose a bit of color discrimination, but I don't really know it.

I will do some tests in the next few days.

renne commented:
@Staars 
I do interpret the linear diagram of the raw lux value like you. RGB seems to be much more complicated. I'm still waiting for an answer from the Broadcom support. I'll be offline for the weekend but try to test your APDS-9960 code on monday.

Staars commented:
Okay, no need to hurry.
I will have to do another PR anyway, because Jason8266 was a bit on the quick side ;)

As I understand the Adafruit-library, they use formulas which can be found in several online-articles to calculate things from e.g. RGB-images.
We will see ...

Happy weekend!


> Am 01.09.2018 um 17:03 schrieb renne <notifications@github.com>:
> 
> @Staars
> I do interpret the linear diagram of the raw lux value like you. RGB seems to be much more complicated. I'm still waiting for an answer from the Broadcom support. I'll be offline for the weekend but try to test your APDS-9960 code on monday.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
